esmvaltool/diag_scripts/kcs/local_resampling.py

```diff
@@ -0,0 +1,465 @@
+from itertools import product
+from pathlib import Path
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import xarray as xr
+from tqdm import tqdm
+
+from esmvaltool.diag_scripts.shared import (ProvenanceLogger,
+                                            get_diagnostic_filename,
+                                            get_plot_filename, run_diagnostic,
+                                            select_metadata)
+
+LOGGER = logging.getLogger(Path(__file__).name)
+
+
+def _create_provenance_record(ancestor_files):
+    """Create a provenance record."""
+    record = {
+        'caption': "Resampling of local climate model.",
+        'domains': ['global'],
+        'authors': [
+            'kalverla_peter',
+            'alidoost_sarah',
+            'rol_evert',
+        ],
+        'ancestors': ancestor_files,
+    }
+    return record
+
+
+def _get_data_target_model(cfg):
+    """Load ensembles for each variable and merge."""
+    LOGGER.info("Reading input data for target model")
+    dataset_dicts = cfg['input_data'].values()
+    dataset = []
+    ancestor_files = []
+    for short_name in "pr", "tas":
+        group = f'{short_name}_target'
+        var = select_metadata(dataset_dicts, variable_group=group)
+        files = [metadata['filename'] for metadata in var]
+        dataset.append(
+            xr.open_mfdataset(files,
+                              concat_dim='ensemble_member',
+                              combine='nested'))
+        ancestor_files.extend(files)
+    provenance = _create_provenance_record(ancestor_files)
+    return xr.merge(dataset).load(), provenance
+
+
+def _segment(dataset, period, step=5):
+    """Compute season means for each n-year segment of input dataset."""
+    segments = []
+    for year in range(*period, step):
+        segments.append(
+            dataset.sel(time=slice(str(year), str(year + step - 1))))
+    segmented_dataset = xr.concat(segments, dim='segment')
+    return segmented_dataset
+
+
+def get_segment_season_means(cfg):
+    """Return a dict with segment season means for control and all futures.
+
+    Read input data for the target model; extract segmented subsets for both
+    the control and future periods; precompute seasonal means for each segment.
+
+    Store intermediate results; if files already exist, load them instead.
+    """
+    dataset, provenance = _get_data_target_model(cfg)
+
+    # Combine the future scenarios (name and resampling period)
+    # with the control period in a single dictionary
+    periods = {
+        name: info['resampling_period']
+        for name, info in cfg['scenarios'].items()
+    }
+    periods['control'] = cfg['control_period']
+
+    # Get the segment season means for each of these periods
+    segments_season_means = {}
+    for name, period in periods.items():
+        LOGGER.info("Get segment season means for %s", name)
+        filename = get_diagnostic_filename(f'season_means_{name}', cfg)
```

Comment:
Is this only an intermediate result or also a final result? Because if it's only intermediate, `cfg['run_dir']` would be a good place to store it.

Reply:
Strictly speaking it is intermediate, but I can image people might want to inspect this file after the thing finishes.

Reply:
In that case, would it make sense to implement provenance for it?

Reply:
I'd say that's a bit overkill. But if you have a strong opinion on this, I'd prefer to change the directory and treat it as intermediate file.

Reply:
Let's do that then, because I think that in the future we want to log a warning or implement a unit test to make sure that provenance is implemented for all output files.

Reply:
Okay @bouweandela I made some modification to make sure the -f and -i flags also work on the run_dir, as we discussed. Can you have a look?